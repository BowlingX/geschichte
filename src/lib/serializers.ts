export interface Serializer {
  readonly deserialize: (value?: string) => any
  readonly serialize: (value?: any) => string
}

const numberSerializer: Serializer = {
  deserialize: (value?: string): number => Number(value),
  serialize: (value?: number): string => String(value)
}

const stringSerializer: Serializer = {
  deserialize: (value?: string): string => (value ? String(value) : ''),
  serialize: (value?: string): string => String(value)
}

export const serializers = {
  number: numberSerializer,
  string: stringSerializer
}
