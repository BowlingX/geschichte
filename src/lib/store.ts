/* tslint:disable:no-expression-statement readonly-array no-shadowed-variable */
import { Draft, enablePatches, produce } from 'immer'
import memoizeOne from 'memoize-one'
import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react'
import create, {
  GetState,
  Mutate,
  SetState,
  State,
  StateCreator,
  StoreApi,
  UseBoundStore,
} from 'zustand'
// tslint:disable-next-line:no-submodule-imports
import { devtools, subscribeWithSelector } from 'zustand/middleware'

// tslint:disable-next-line:no-submodule-imports
import shallow from 'zustand/shallow'
import {
  converter,
  historyManagement,
  immerWithPatches,
  StoreState,
} from './middleware'
import { Serializer } from './serializers'
import {
  applyFlatConfigToState,
  createQueryObject,
  flattenConfig,
} from './utils'

enablePatches()

export const DEFAULT_NAMESPACE = 'default'
export const StoreContext = createContext<StoreApi<StoreState<any>> | null>(
  null
)

export interface Parameter<V = any> {
  readonly name: string
  readonly serializer: Serializer<V>
  // tslint:disable-next-line:no-mixed-interface
  readonly skipValue: (value?: V, initialValue?: V) => boolean
}

export interface MappedParameter extends Parameter {
  readonly path: readonly string[]
}

export interface Config {
  readonly [propName: string]: Config | (() => Parameter)
}

export interface MappedConfig {
  readonly [queryParameter: string]: MappedParameter
}

export type RouterOptions = Record<string, any>

export interface HistoryManagement {
  /** the initial search string (e.g. ?query=test), contains the questionsmark */
  readonly initialSearch: () => string
  readonly push: (
    queryObject: Record<string, string>,
    options?: RouterOptions
  ) => Promise<unknown>
  readonly replace: (
    queryObject: Record<string, string>,
    options?: RouterOptions
  ) => Promise<unknown>
}

export const useGeschichte = <T extends State>(
  historyInstance: HistoryManagement
) => {
  const thisStore = converter<T>(historyInstance)
  const storeWithHistory = historyManagement<T>(historyInstance)(thisStore)

  const middleware = immerWithPatches(
    storeWithHistory
  ) as unknown as StateCreator<StoreState<T>, any>

  if (process.env.NODE_ENV === 'development' && typeof window !== 'undefined') {
    return create(
      subscribeWithSelector(devtools(middleware, { name: 'geschichte' }))
    )
  }
  return create(subscribeWithSelector(middleware))
}

type InitialValuesProvider<T> = T | (() => T)

export const useStore = <T extends object>() => {
  return useContext(StoreContext) as UseBoundStore<StoreApi<StoreState<T>>>
}

export const useBatchQuery = <T extends State>() => {
  const store = useStore<T>()
  return store(
    ({ batchPushState, batchReplaceState }) => ({
      batchPushState,
      batchReplaceState,
    }),
    shallow
  )
}

export const factoryParameters = <T extends object>(
  config: Config,
  // tslint:disable-next-line:no-object-literal-type-assertion
  defaultInitialValues: InitialValuesProvider<T> = {} as T,
  ns: string = DEFAULT_NAMESPACE
) => {
  const flatConfig = flattenConfig(config)
  const createInitialValues = (d: InitialValuesProvider<T>) =>
    typeof d === 'function' ? (d as () => T)() : d

  const initBlank = (initialQueries: object, initialValues: T) => {
    // thisValues will be mutated by applyFlatConfigToState, that's why we init it with a copy of
    // the initial state.
    // tslint:disable-next-line:no-let
    let thisQuery = {}
    // We produce a new state here instead of mutating defaultInitialValues.
    // Otherwise it's possible that it get's reused across executions and that will yield to readonly errors.
    const values = produce(initialValues, (draft: Draft<T>) => {
      thisQuery = applyFlatConfigToState(
        flatConfig,
        initialQueries,
        ns,
        draft as T,
        initialValues as T
      )
    })
    return {
      initialValues,
      query: thisQuery,
      values,
    }
  }

  const memInitBlank = memoizeOne(initBlank)
  const memCreateInitialValues = memoizeOne(createInitialValues)

  const useQuery = () => {
    const useStore = useContext(StoreContext) as UseBoundStore<
      Mutate<
        StoreApi<StoreState<T>>,
        [['zustand/subscribeWithSelector', never]]
      >
    >

    const {
      register,
      pushState,
      replaceState,
      resetPush,
      resetReplace,
      initialQueries,
    } = useStore(
      ({
        register,
        pushState,
        replaceState,
        resetPush,
        resetReplace,
        initialQueries,
      }) => ({
        initialQueries,
        pushState,
        register,
        replaceState,
        resetPush,
        resetReplace,
      }),
      shallow
    )
    const initialRegisterState = useMemo(() => {
      const initialValues = memCreateInitialValues(defaultInitialValues)
      return memInitBlank(initialQueries(), initialValues)
    }, [useStore, defaultInitialValues])

    const [currentState, setCurrentState] = useState({
      initialValues: initialRegisterState.initialValues,
      values: initialRegisterState.values,
    })

    useEffect(() => {
      const {
        unsubscribe: unregister,
        values,
        initialValues,
      } = register(
        config,
        flatConfig,
        ns,
        initialRegisterState.initialValues,
        initialRegisterState.query,
        initialRegisterState.values
      )
      if (
        initialRegisterState.values !== values ||
        initialRegisterState.initialValues !== initialValues ||
        currentState.initialValues !== initialValues ||
        currentState.values !== values
      ) {
        setCurrentState({ values, initialValues })
      }
      const unsubscribe = useStore.subscribe(
        (state) =>
          state.namespaces[ns] && {
            initialValues: state.namespaces[ns].initialValues,
            values: state.namespaces[ns].values,
          },
        (state) => {
          if (state) {
            setCurrentState(state)
          }
        },
        { equalityFn: shallow }
      )

      return () => {
        unsubscribe()
        unregister()
      }
    }, [initialRegisterState])

    const values = currentState.values
    const initialValues = currentState.initialValues

    const createQuery = useCallback(
      (values: Partial<T>) => {
        return createQueryObject(flatConfig, ns, values, initialValues)
      },
      [initialValues]
    )

    return useMemo(
      () => ({
        createQuery: (customValues?: Partial<T>) =>
          createQuery(customValues || values),
        createQueryString: (customValues?: Partial<T>) =>
          new URLSearchParams(createQuery(customValues || values)).toString(),
        initialValues,
        pushState: (state: (state: T) => void, options?: Record<string, any>) =>
          pushState(ns, state, options),
        replaceState: (
          state: (state: T) => void,
          options?: Record<string, any>
        ) => replaceState(ns, state, options),
        resetPush: () => resetPush(ns),
        resetReplace: () => resetReplace(ns),
        values,
      }),
      [
        values,
        initialValues,
        pushState,
        replaceState,
        resetPush,
        resetReplace,
        createQuery,
      ]
    )
  }

  const createQueryString = (
    values: Partial<T>,
    initialValues?: Partial<T> | null
  ): string => {
    const thisInitialValues =
      typeof initialValues === 'undefined'
        ? memCreateInitialValues(defaultInitialValues)
        : initialValues
    return new URLSearchParams(
      createQueryObject<T>(flatConfig, ns, values, thisInitialValues)
    ).toString()
  }

  const parseQueryString = (
    query: string,
    initialValues?: Partial<T> | null
  ): Partial<T> => {
    const thisInitialValues =
      typeof initialValues === 'undefined'
        ? memCreateInitialValues(defaultInitialValues)
        : initialValues
    return produce({}, (draft: Draft<T>) => {
      const parsedQuery = new URLSearchParams(query)
      applyFlatConfigToState(
        flatConfig,
        Object.fromEntries(parsedQuery.entries()),
        ns,
        draft as T,
        thisInitialValues as T
      )
    })
  }

  return { useQuery, createQueryString, parseQueryString }
}
