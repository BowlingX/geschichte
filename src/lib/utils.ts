/* tslint:disable:no-expression-statement no-let no-submodule-imports no-object-mutation */
import { Patch } from 'immer'
import shallowEqual from 'zustand/shallow'
import { GenericObject } from './middleware'
import { Serializer } from './serializers'
import { Config, DEFAULT_NAMESPACE, MappedConfig, Parameter } from './store'

export const pm = (name: string, serializer: Serializer) => (): Parameter => ({
  name,
  serializer
})

export const createOrApplyPath = (
  obj: GenericObject | null,
  path: readonly string[],
  value: any = null
) => {
  let current = obj || {}
  let thisPath: ReadonlyArray<string> = [...path]
  while (thisPath.length > 1) {
    const [head, ...tail] = thisPath
    thisPath = tail
    if (current[head] === undefined) {
      current[head] = {}
    }
    current = current[head]
  }
  current[thisPath[0]] = value
  return obj
}

export const formatNamespace = (key: string, ns?: string) => {
  return ns && ns !== DEFAULT_NAMESPACE ? `${ns}.${key}` : key
}

export const get = <T = object>(
  object: T,
  path: ReadonlyArray<string | number>
) => {
  return path.reduce((next: T | any, key: string | number) => {
    return next ? next[key] : undefined
  }, object)
}

/**
 * Default skip implementation
 */
export const skipValue = (value?: any, initialValue?: any) =>
  value === undefined ||
  value === null ||
  shallowEqual(value, initialValue) ||
  (Array.isArray(value) && value.length === 0)

/**
 * @return a list of patches that deeply match the given config object
 */
const findDeepPatches = (
  config: Config,
  basePath: readonly string[]
): readonly Patch[] => {
  return Object.keys(config).reduce((next: readonly any[], item: string) => {
    if (typeof config[item] === 'function') {
      return [...next, { path: [...basePath, item], op: 'replace' }]
    }
    return [
      ...next,
      ...findDeepPatches(config[item] as Config, [...basePath, item])
    ]
  }, [])
}

/**
 * @return an object with the keys that have been processed
 * if a key has been removed / set to undefined, we still return it to
 * be able to create a diff to the current state
 */
export const createQueriesFromPatch = <T = object>(
  config: Config,
  ns: string,
  patch: readonly Patch[],
  state: T,
  initialState: T
): object => {
  return patch.reduce((next, item) => {
    const { path, op } = item
    // namespaces, [ns], values|initialValues, ...rest
    const [, patchNamespace, , ...objectPath] = path
    // skip patches that don't belong to the given namespace
    if (patchNamespace !== ns) {
      return next
    }
    const possibleParameter = get(config, objectPath)
    const isNotCallable = typeof possibleParameter !== 'function'

    if (possibleParameter !== undefined && isNotCallable) {
      // If we have an object as result, we create patches for each parameter inside the subtree
      const patches = findDeepPatches(
        possibleParameter,
        path as readonly string[]
      )
      return {
        ...next,
        ...createQueriesFromPatch<T>(config, ns, patches, state, initialState)
      }
    }

    if (isNotCallable) {
      return next
    }

    const { name, serializer, skip } = possibleParameter()
    // @ts-ignore
    const value = get(state, objectPath)
    // @ts-ignore
    const initialValue = get(initialState, objectPath)

    const nextValue = skipValue(value, initialValue) ? undefined : value

    return {
      ...next,
      [formatNamespace(name, ns)]:
        nextValue === undefined ? nextValue : serializer.serialize(nextValue)
    }
  }, {})
}

/**
 * Creates a queryObject that can be serialized.
 */
export const createQueryObject = <T = object>(
  config: MappedConfig,
  ns: string,
  values: Partial<T>,
  initialState?: Partial<T> | null
) => {
  return Object.keys(config).reduce((next, parameter) => {
    const { path, serializer } = config[parameter]
    const possibleValue = get(values, path)
    const nextValue = skipValue(possibleValue, get(initialState, path))
      ? undefined
      : possibleValue
    if (nextValue === undefined) {
      return next
    }
    return {
      ...next,
      [formatNamespace(parameter, ns)]: serializer.serialize(nextValue)
    }
  }, {})
}

export const applyDiffWithCreateQueriesFromPatch = <T = object>(
  config: Config,
  ns: string,
  currentQuery: object,
  patch: readonly Patch[],
  state: T,
  initialState: T
) => {
  const query = createQueriesFromPatch(config, ns, patch, state, initialState)
  const nextQueries: GenericObject = {
    ...currentQuery,
    ...query
  }

  return Object.keys(nextQueries)
    .filter(key => nextQueries[key] !== undefined)
    .reduce((next, key) => {
      return {
        ...next,
        [key]: nextQueries[key]
      }
    }, {})
}

/**
 * Applies the given queryValues to the current state, based on the configuration
 * Important: Mutates `state`.
 * @return an object with the keys that have been processed
 */
export const applyFlatConfigToState = <T = object>(
  config: MappedConfig,
  queryValues: { readonly [index: string]: any },
  ns: string,
  state: T,
  initialState: T
) => {
  return Object.keys(config).reduce((next, queryParameter) => {
    const { path, serializer } = config[queryParameter]
    const nsQueryParameter = formatNamespace(queryParameter, ns)
    const maybeValue = queryValues[nsQueryParameter]

    const value =
      maybeValue === undefined
        ? get(initialState, path)
        : serializer.deserialize(maybeValue)
    createOrApplyPath(state, path, value)

    if (skipValue(value, get(initialState, path))) {
      return next
    }

    return {
      ...next,
      [nsQueryParameter]: maybeValue
    }
  }, {})
}

/**
 * Creates a flat serializable object based on a mapping Object
 * that can be used to generate a query string.
 * @return a key/value object with the object containing the defined parameters as key.
 */
export const flattenConfig = (
  config: Config,
  path: readonly string[] = []
): MappedConfig => {
  return Object.keys(config).reduce(
    (next: { readonly [index: string]: any }, key) => {
      const v = config[key]
      const nextPath: ReadonlyArray<string> = [...path, key]
      if (typeof v === 'function') {
        const { name, ...rest } = v()
        if (next[name] !== undefined) {
          throw new Error(
            `Config invalid: Multiple definitions found for ${name}.`
          )
        }
        return {
          ...next,
          [name]: {
            path: nextPath,
            ...rest
          }
        }
      }
      if (typeof v === 'object') {
        return {
          ...next,
          ...flattenConfig(v, nextPath)
        }
      }
      return next
    },
    {}
  )
}
