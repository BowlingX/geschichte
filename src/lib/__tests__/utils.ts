/* tslint:disable:no-expression-statement */
import expect from 'expect'
import { Patch } from 'immer'
import { serializers } from '../serializers'
import { DEFAULT_NAMESPACE } from '../store'
import { createQueriesFromPatch, flattenConfig, pm } from '../utils'

describe('utils', () => {
  describe('flattenConfig', () => {
    const config = {
      somethingElse: {
        deeper: pm('anotherOne', serializers.int),
        evenDeeper: {
          test: pm('deeper', serializers.string)
        }
      },
      state: pm('theParameter', serializers.int)
    }
    it('should create a flat config from `config`', () => {
      const flatConfig = flattenConfig(config)
      expect(Object.keys(flatConfig)).toEqual([
        'anotherOne',
        'deeper',
        'theParameter'
      ])
      expect(flatConfig).toMatchSnapshot()
    })
  })

  describe('createQueriesFromPatch', () => {
    const config = {
      parameter: pm('p', serializers.string)
    }
    it('should generate undefined values for undefined queries', () => {
      const patch: readonly Patch[] = [
        {
          op: 'add',
          path: ['namespaces', DEFAULT_NAMESPACE, 'values', 'parameter'],
          value: undefined
        }
      ]
      const nextQueries = createQueriesFromPatch(
        config,
        DEFAULT_NAMESPACE,
        patch,
        { parameter: undefined },
        { parameter: 'something' }
      )
      expect(nextQueries).toEqual({ p: undefined })
    })
  })
})
