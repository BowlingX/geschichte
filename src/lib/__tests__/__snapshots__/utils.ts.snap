// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils flattenConfig should create a flat config from \`config\` 1`] = `
{
  "anotherOne": {
    "path": [
      "somethingElse",
      "deeper",
    ],
    "serializer": {
      "deserialize": [Function],
      "serialize": [Function],
    },
    "skipValue": [Function],
  },
  "deeper": {
    "path": [
      "somethingElse",
      "evenDeeper",
      "test",
    ],
    "serializer": {
      "deserialize": [Function],
      "serialize": [Function],
    },
    "skipValue": [Function],
  },
  "theParameter": {
    "path": [
      "state",
    ],
    "serializer": {
      "deserialize": [Function],
      "serialize": [Function],
    },
    "skipValue": [Function],
  },
}
`;
